File: packages/modules/ImsMedia/service/src/com/android/telephony/imsmedia/lib/libimsmedia/config/src/MediaQualityStatus.cpp
/**
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <MediaQualityStatus.h>

namespace android
{

namespace telephony
{

namespace imsmedia
{

MediaQualityStatus::MediaQualityStatus()
{
    mRtpInactivityTimeMillis = 0;
    mRtcpInactivityTimeMillis = 0;
    mRtpPacketLossRate = 0;
    mRtpJitterMillis = 0;
}

MediaQualityStatus::MediaQualityStatus(const MediaQualityStatus& status)
{
    mRtpInactivityTimeMillis = status.mRtpInactivityTimeMillis;
    mRtcpInactivityTimeMillis = status.mRtcpInactivityTimeMillis;
    mRtpPacketLossRate = status.mRtpPacketLossRate;
    mRtpJitterMillis = status.mRtpJitterMillis;
}

MediaQualityStatus::~MediaQualityStatus() {}

MediaQualityStatus& MediaQualityStatus::operator=(const MediaQualityStatus& status)
{
    if (this != &status)
    {
        mRtpInactivityTimeMillis = status.mRtpInactivityTimeMillis;
        mRtcpInactivityTimeMillis = status.mRtcpInactivityTimeMillis;
        mRtpPacketLossRate = status.mRtpPacketLossRate;
        mRtpJitterMillis = status.mRtpJitterMillis;
    }

    return *this;
}

bool MediaQualityStatus::operator==(const MediaQualityStatus& status) const
{
    return (mRtpInactivityTimeMillis == status.mRtpInactivityTimeMillis &&
            mRtcpInactivityTimeMillis == status.mRtcpInactivityTimeMillis &&
            mRtpPacketLossRate == status.mRtpPacketLossRate &&
            mRtpJitterMillis == status.mRtpJitterMillis);
}

bool MediaQualityStatus::operator!=(const MediaQualityStatus& status) const
{
    return (mRtpInactivityTimeMillis != status.mRtpInactivityTimeMillis ||
            mRtcpInactivityTimeMillis != status.mRtcpInactivityTimeMillis ||
            mRtpPacketLossRate != status.mRtpPacketLossRate ||
            mRtpJitterMillis != status.mRtpJitterMillis);
}

status_t MediaQualityStatus::writeToParcel(Parcel* out) const
{
    out->writeInt32(mRtpInactivityTimeMillis);
    out->writeInt32(mRtcpInactivityTimeMillis);
    out->writeInt32(mRtpPacketLossRate);
    out->writeInt32(mRtpJitterMillis);
    return NO_ERROR;
}

status_t MediaQualityStatus::readFromParcel(const Parcel* in)
{
    in->readInt32(&mRtpInactivityTimeMillis);
    in->readInt32(&mRtcpInactivityTimeMillis);
    in->readInt32(&mRtpPacketLossRate);
    in->readInt32(&mRtpJitterMillis);
    return NO_ERROR;
}

void MediaQualityStatus::setRtpInactivityTimeMillis(int32_t time)
{
    mRtpInactivityTimeMillis = time;
}

int32_t MediaQualityStatus::getRtpInactivityTimeMillis()
{
    return mRtpInactivityTimeMillis;
}

void MediaQualityStatus::setRtcpInactivityTimeMillis(int32_t time)
{
    mRtcpInactivityTimeMillis = time;
}

int32_t MediaQualityStatus::getRtcpInactivityTimeMillis()
{
    return mRtcpInactivityTimeMillis;
}

void MediaQualityStatus::setRtpPacketLossRate(int32_t rate)
{
    mRtpPacketLossRate = rate;
}

int32_t MediaQualityStatus::getRtpPacketLossRate()
{
    return mRtpPacketLossRate;
}

void MediaQualityStatus::setRtpJitterMillis(int32_t jitter)
{
    mRtpJitterMillis = jitter;
}

int32_t MediaQualityStatus::getRtpJitterMillis()
{
    return mRtpJitterMillis;
}

}  // namespace imsmedia

}  // namespace telephony

}  // namespace android

File: packages/modules/ImsMedia/framework/src/android/telephony/imsmedia/MediaQualityStatus.java
/**
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.telephony.imsmedia;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Objects;

/**
 * Class to set the media quality status for notifications
 *
 * @hide
 */

public final class MediaQualityStatus implements Parcelable {
    private final int mRtpInactivityTimeMillis;
    private final int mRtcpInactivityTimeMillis;
    private final int mRtpPacketLossRate;
    private final int mRtpJitterMillis;

    /** @hide **/
    public MediaQualityStatus(Parcel in) {
        mRtpInactivityTimeMillis = in.readInt();
        mRtcpInactivityTimeMillis = in.readInt();
        mRtpPacketLossRate = in.readInt();
        mRtpJitterMillis = in.readInt();
    }

    /** @hide **/
    public MediaQualityStatus(Builder builder) {
        mRtpInactivityTimeMillis = builder.mRtpInactivityTimeMillis;
        mRtcpInactivityTimeMillis = builder.mRtcpInactivityTimeMillis;
        mRtpPacketLossRate = builder.mRtpPacketLossRate;
        mRtpJitterMillis = builder.mRtpJitterMillis;
    }

    /** @hide **/
    public int getRtpInactivityTimeMillis() {
        return mRtpInactivityTimeMillis;
    }

    /** @hide **/
    public int getRtcpInactivityTimeMillis() {
        return mRtcpInactivityTimeMillis;
    }

    /** @hide **/
    public int getRtpPacketLossRate() {
        return mRtpPacketLossRate;
    }

    /** @hide **/
    public int getRtpJitterMillis() {
        return mRtpJitterMillis;
    }

    @NonNull
    @Override
    public String toString() {
        return "MediaQualityStatus: {mRtpInactivityTimeMillis=" + mRtpInactivityTimeMillis
                + ", mRtcpInactivityTimeMillis=" + mRtcpInactivityTimeMillis
                + ", mRtpPacketLossRate=" + mRtpPacketLossRate
                + ", mRtpJitterMillis=" + mRtpJitterMillis
                + " }";
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                mRtpInactivityTimeMillis, mRtcpInactivityTimeMillis, mRtpPacketLossRate,
                mRtpJitterMillis);
    }

    @Override
    public boolean equals(@Nullable Object o) {
        if (o == null || !(o instanceof MediaQualityStatus) || hashCode() != o.hashCode()) {
            return false;
        }

        if (this == o) {
            return true;
        }

        MediaQualityStatus s = (MediaQualityStatus) o;

        return (mRtpInactivityTimeMillis == s.mRtpInactivityTimeMillis
                && mRtcpInactivityTimeMillis == s.mRtcpInactivityTimeMillis
                && mRtpPacketLossRate == s.mRtpPacketLossRate
                && mRtpJitterMillis == s.mRtpJitterMillis);
    }

    /**
     * {@link Parcelable#describeContents}
     */
    public int describeContents() {
        return 0;
    }

    /**
     * {@link Parcelable#writeToParcel}
     */
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(mRtpInactivityTimeMillis);
        dest.writeInt(mRtcpInactivityTimeMillis);
        dest.writeInt(mRtpPacketLossRate);
        dest.writeInt(mRtpJitterMillis);
    }

    public static final @NonNull Parcelable.Creator<MediaQualityStatus>
                CREATOR = new Parcelable.Creator() {
                    public MediaQualityStatus createFromParcel(Parcel in) {
                            // TODO use builder class so it will validate
                            return new MediaQualityStatus(in);
                    }

                    public MediaQualityStatus[] newArray(int size) {
                            return new MediaQualityStatus[size];
                    }
                };

    /**
     * Provides a convenient way to set the fields of a {@link MediaQualityStatus}
     * when creating a new instance.
     */
    public static final class Builder {
        private int mRtpInactivityTimeMillis;
        private int mRtcpInactivityTimeMillis;
        private int mRtpPacketLossRate;
        private int mRtpJitterMillis;

        /**
         * Set the rtp inactivity observed as per thresholds set by the MediaQualityThreshold API
         * @param value The receiving rtp inacitivity time observed in milliseconds unit
         */
        public @NonNull Builder setRtpInactivityTimeMillis(int value) {
            this.mRtpInactivityTimeMillis = value;
            return this;
        }

        /**
         * Set the rtcp inactivity observed as per thresholds set by the MediaQualityThreshold API
         * @param value The receiving rtcp inacitivity time observed in milliseconds unit
         */
        public @NonNull Builder setRtcpInactivityTimeMillis(int value) {
            this.mRtcpInactivityTimeMillis = value;
            return this;
        }

        /**
         * Set the rtp packet loss rate observed as per thresholds set by the MediaQualityThreshold
         * API
         * @param value The receiving rtp packet loss rate calculated in percentage unit
         */
        public @NonNull Builder setRtpPacketLossRate(int value) {
            this.mRtpPacketLossRate = value;
            return this;
        }

        /**
         * Set the rtp jitter observed as per thresholds set by MediaQualityThreshold API
         * @param value The receiving rtp jitter calculated in milliseconds unit
         */
        public @NonNull Builder setRtpJitterMillis(int value) {
            this.mRtpJitterMillis = value;
            return this;
        }

        /**
         * Build the MediaQualityStatus.
         *
         * @return the MediaQualityStatus object.
         */
        public @NonNull MediaQualityStatus build() {
            // TODO validation
            return new MediaQualityStatus(this);
        }
    }
}
